Default values for struct fields

KodrAus opened the pull request:
RFC content for #1 

cc: @nrc @leodasvacas

KodrAus committed:
initial thoughts on default fields

KodrAus commented on default-fields/0000-default-fields.md:
```diff
@@ -11,14 +11,182 @@ One para explanation of the feature.
+If it's unclear whether or not a particular caller will use a default field value then
+its addition can't be treated as a non-breaking change.
+
+The semantics of field defaults are the same as consts.
```
Happy to be convinced otherwise on this one, but following the same semantics as `const` seemed like a good starting point.

Mark-Simulacrum commented on default-fields/0000-default-fields.md:
```diff
@@ -6,19 +6,224 @@
+#[derive(Default)]
+struct Foo {
+    a: &'static str,
+    b: bool = true
```
nit: "," at the end of the line

Mark-Simulacrum commented on default-fields/0000-default-fields.md:
````diff
@@ -6,19 +6,224 @@
+```rust
+impl Default for Foo {
+	fn default() -> Self {
+		Foo {
````
nit: Indentation here looks a little off

nrc replied:
prefer spaces to tabs

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,19 +6,224 @@
+```
+
+This requires all fields in the struct be assigned a value, unless the struct implements
+the Default trait.
````
This doesn't rely on the Default trait, you can supply any struct instance after the `..` and the new struct takes its values.

KodrAus replied:
I didn't actually know that :+1: That's pretty neat

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,19 +6,224 @@
+The problem with this is that it requires Foo to derive Default, which may not always
+be desirable.
+
+The existing struct construction also can't be used when a struct has private fields:
```
prefer "initialisation" to "construction"

KodrAus replied:
:white_check_mark:

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,19 +6,224 @@
+
+To work around these shortcomings, users can create constructor functions or more elaborate builders.
+The problem with a constructor is that it needs to be kept in sync with the fields of the struct, or
+require the user reassign values they want to change.
```
I'd say a bigger problem with ctors is you need one for each combination of fields

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,19 +6,224 @@
+Builders enable more advanced construction, but also need to kept in-line with the struct they build,
+and can be overkill for large, but uncomplicated structs.
+
+With field defaults a caller can instantiate a struct without needing builders
```
From here on down is presenting the feature rather than motivating it, so I would move it to the next section

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,19 +6,224 @@
+    b: true,
+    c: 42
+};
+```
````
I would also mention that for big structs which always get the same value, it is repetitive and inconvenient to always instantiate each field.

KodrAus replied:
I would think that case is easily solved by a constructor function

nrc replied:
true, but it is still annoying boiler-plate even if it is just in a ctor function, and it is not uncommon to have multiple ctor functions

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,19 +6,224 @@
 
-Why are we doing this? What use cases does it support? What is the expected outcome?
+Today, Rust allows you to create an instance of a struct using a literal syntax:
+
```
structure-wise, I would give short sentences for each of the pieces of motivation here, then explain in more detail like you are doing.

I don't think you need this first example

leoyvens replied:
Is there a standard terminology for this syntax? Some languages call them "struct literals", maybe we can call them "struct initialisers", the reference uses "struct expressions".

nrc replied:
I usually call them struct literals, I don't think there is standard terminology.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,19 +6,224 @@
-This is the bulk of the RFC. Explain the design in enough detail for somebody familiar
-with the language to understand, and for somebody familiar with the compiler to implement.
-This should get into specifics and corner-cases, and include examples of how the feature is used.
+The semantics of field defaults are the same as consts.
```
This sentence is a bit vague - semantics could mean a lot of things. I would make this a note that we are modelling field defaults on `const`s and move it later.

I would start with the syntax, then explain restrictions on the value.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,19 +6,224 @@
+The semantics of field defaults are the same as consts.
+The type must be supplied, and the value must be a valid compile-time constant.
+
+Field defaults take precedence over `Default::default()`.
```
Make clear that this is a feature of derive(Default)

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,19 +6,224 @@
+		}
+	}
+}
+```
````
Probably don't need this expansion.

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,19 +6,224 @@
+}
+```
+
+Supplied field values take precedence over field defaults:
````
I would move this up above the derive(Default) stuff

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,19 +6,224 @@
+};
+```
+
+That means if `Default` is manually implemented for a struct and default fields are
````
I don't think you need this section on Default. On the other hand, I would add some detail on function update syntax.

KodrAus replied:
Could you explain what you mean by _function update syntax_? You've lost me on that one :)

leoyvens replied:
function update syntax or functional record update is the formal name for `..Foo }` syntax.

nrc replied:
yeah sorry, I meant functional, not function.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -28,9 +233,29 @@ Why should we *not* do this?
 [alternatives]: #alternatives
 
-What other designs have been considered? What is the impact of not doing this?
+Just stick with builders and constructor functions.
```
This is basically just saying "don't do this" which is not very useful. It would be better to list tweaks to the design, and explain why our proposal is better than those alternatives

KodrAus replied:
Yeh, this is basically a placeholder. It turns out writing about alternatives and drawbacks is kind of difficult.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -28,9 +233,29 @@ Why should we *not* do this?
+Is `const` the right feature to base off?
+It means field values are just as constrained as `const` is now,
+but also prevents struct definitions from becomes bloated by complex logic that
+maybe belongs in a builder instead.
```
I would move this to alternatives (and phrased a bit differently). Also consider runtime overhead/expectations of struct literal vs function call

KodrAus replied:
I'm investigating this one now. It would depend on the way this feature is implemented right?

nrc replied:
yes

KodrAus replied:
Is there anywhere I can get a sense of how struct literals work now, so I've got some context of how this should be implemented?

I think the simplest implementation would be to treat struct initialisation with field defaults as sugar for the 'real' initialiser, where defaults for missing fields are spliced in.

In this case, assuming no inlining, you aren't paying for a function call. Because this is opaque to the caller, with field defaults there might be other optimisations the compiler could do to make it more efficient. I'm not sure what those could be yet though.

nrc replied:
So, actually, let me qualify my earlier "yes". I think it depends more on the design of the details of the feature than on the actual implementation.

If you want to investigate the details of how struct lits are compiled, you need to look at the compiler, or possibly dump the LLVM output. But really, there is nothing surprising happening atm in compilation. The only detail really FRU.

What is important, I think, is that there is no initialisation at all, it really is just a literal expression. That is, it just copies the field values into a given bit of memory.

> I think the simplest implementation would be to treat struct initialisation with field defaults as sugar for the 'real' initialiser, where defaults for missing fields are spliced in.

I agree, and using const values for default values fits well with this model - there is no extra runtime overhead at all. Anything beyond const values (I think) introduces some runtime overhead. If you allow `Default` integration or other extensions, then there is runtime overhead.

So, I agree with your original points - "complex logic that maybe belongs in a builder instead", and I'm trying to dig into the 'why' behind this point. I think expectations around runtime overhead are part of that.

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -28,9 +233,29 @@ Why should we *not* do this?
+}
+```
+
+Should this type automatically derive `Default`?
````
Generally, we always opt-in to non-auto-traits.

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -28,9 +233,29 @@ Why should we *not* do this?
+```
+
+Should this type automatically derive `Default`?
+Should we be able to construct this type as if it's a ZST, like `let foo = Foo;`?
````
This is a unit struct, you can also use `Foo {}` for an empty struct, and the braces must match - you can't use one declaration with the other construction. That suggests you can't use `Foo`, but need `Foo{}` (also note that the two forms do have slightly different rules for naming).

KodrAus replied:
That makes sense :+1:

KodrAus committed:
address some feedback

KodrAus commented:
I'll work through this feedback and give it a good complete pass over this weekend. I'll shoot a ping when I'm finished.

KodrAus committed:
more work on feedback

KodrAus committed:
work on alternatives section

KodrAus commented on default-fields/0000-default-fields.md:
````diff
@@ -6,19 +6,224 @@
+```rust
+impl Default for Foo {
+	fn default() -> Self {
+		Foo {
````
:white_check_mark:

KodrAus committed:
clean up later sections

Commit:
elaborate expectations

KodrAus commented:
@nrc Sorry it's taken so long to get to this point, but I'm now happy with the state of this. I've addressed the comments thus far and think it's ready for another round of feedback.

leoyvens commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,201 @@
+    a: "Hello",
+    c: 42,
+}
+```
````
Would be good to add a field using `default()`.

leoyvens commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,201 @@
+Today, Rust allows you to create an instance of a struct using a literal syntax.
+This requires all fields in the struct be assigned a value, and can't be used
+with private fields.
+It can also be inconvenient for large structs that usually receive the same values.
```
This sentence sounds like the entire struct receive the same value, maybe reword to clarify that the fields receive the same value.

leoyvens commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,201 @@
+
+With the `..` syntax, values for missing fields can be taken from another struct.
+This is convenient for using `Default` to override just a few values on initialisation:
+
```
"This is convenient for using `Default`" -> "For example using `Default` to override..."

leoyvens commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,201 @@
+}
+```
+
+However, this still requires an already initialised struct after the `..`.
````
Another drawback is that if the struct has a private field `..Default::default()` isn't valid.

KodrAus replied:
There you go. Another thing I didn't know about `..` 👍

nrc replied:
And requires the type of every field to implement `Default`, which is not really common.

leoyvens commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,201 @@
+However, this still requires an already initialised struct after the `..`.
+
+To work around these shortcomings, users can create constructor functions or more elaborate builders.
+The problem with a constructor is that you need one for each combination of fields a caller can supply.
```
Or supply one constructor with mandatory values and expect the user to override public default values after construction.

``` rust
let foo = Foo::new(42); // sets `foo.c`
foo.b = true; // override b
```

leoyvens commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,201 @@
+
+Field defaults use the same syntax as `const`s.
+So the type must be supplied, and the value must be a compile-time expression, or a call to `Default::default()`.
+Default field values are expected to be cheap to produce and have basic support for collection types like `Vec`.
```
Inference is an alternative here, you have to name the type when using `Foo::default()` anyways.

KodrAus replied:
I've opted for explicit typing because it's easy to follow when the types all appear in the same place. It's not a feature I'm particularly attached to though, so if you think it's not _really_ valuable then I'll rejig it.

nrc replied:
I think starting with const syntax is fine. You should be clear that it is the same syntax and set of initialising expressions as consts. I would put the `default` option in a separate paragraph. with its justification.

KodrAus replied:
Note this is a bit different now that default is in alternatives. I've added a point that the syntax and functionality are modeled after `const`. Maybe it could be clearer though, since the syntax is in the grammar section, and the functionality is in interpretation.

Commit:
address some feedback

KodrAus commented:
@leodasvacas I've addressed your feedback except for the point about inferring types for `default()` fields. I thought we could discuss that.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
+    a: &'static str,
+    b: bool = true,
+    c: i32,
+    d: Vec<i32> = Vec::default()
```
do you need the `Vec` here? I'd expect not.

KodrAus replied:
So we're happy with type inference for default expressions? I'd left the type as mandatory to align with `const`, but like I said to @leodasvacas I'm not really attached to it. If you both think inference is the way to go here I'll tweak the relevant sections.

nrc replied:
No I meant on the rhs, sorry,  I would have thought `d: Vec<i32> = default()` should work? Which reminds me, in the detailed design you should specify _how_ the compiler will allow `default()` to work - presumably `Default` is a lang item, and it can check that the rhs is a single function call, which calls the single method of that trait (as opposed to say checking for syntactic equality).

leoyvens replied:
`default()` is not in the prelude so we would need `default()` to be syntatic sugar in struct declarations. That would be assymetrical to normal uses of `default()`, type inference on lhs would be symetrical to how `let` works.

nrc replied:
Hmm, I guess you could use `Default::default()` or `Vec::default()` or even `<Vec as Default>::default()`, but all would require `Default` be imported.

leoyvens replied:
To be totally clear `Default` is in the prelude, however `Default::default()` is not.

KodrAus replied:
Hmm, so we'd have a lang item that looks a little something like (pseudo):

``` rust
struct DefaultField(Ty, DefaultFieldKind)

enum DefaultFieldKind {
    Const(Expr),
    Default
}
```

Where default would expand to `Default::default()`?

nrc replied:
hmm, this feels a bit like overkill. I was thinking more that the `Default` trait could be a lang item. However, all this talk of syntactic sugar for something that looks like a function call makes me nervous. I saw `..` suggested as special syntax for this at some point. However, I'm thinking that is might be best to drop the default special case to start with and just list it as an alternative.

KodrAus replied:
I've moved `default` to alternatives. Is it pretty common for alternatives in an RFC to make their way into the proposal proper?

nrc replied:
It happens, I'm not sure how common it is. Depends a lot on the RFC and the alternative, I think.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
-Why are we doing this? What use cases does it support? What is the expected outcome?
+Today, Rust allows you to create an instance of a struct using a literal syntax.
+This requires all fields in the struct be assigned a value, and can't be used
+with private fields.
```
This is not quite right, you can use the syntax with private fields, if you have access to the private fields, it is only a problem when you do not.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
+It also isn't valid if the struct has private fields.
+
+To work around these shortcomings, users can create constructor functions or more elaborate builders.
+The problem with a constructor is that you either need one for each combination of fields a caller can supply
```
I would assume that people don't do this - either you have one (or a very small number) of constructors, or you use a builder.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
 [motivation]: #motivation
 
-Why are we doing this? What use cases does it support? What is the expected outcome?
+Today, Rust allows you to create an instance of a struct using a literal syntax.
```
I would start this section with a summary para which lists the motivations - privacy and boilerplate and that field defaults address this. Then go into detail about the two issue. ATM, the motivation reads like a few random issues and doesn't relate them to the proposal.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
-This is the bulk of the RFC. Explain the design in enough detail for somebody familiar
-with the language to understand, and for somebody familiar with the compiler to implement.
-This should get into specifics and corner-cases, and include examples of how the feature is used.
+With field defaults a caller can initialise a struct with default values without needing builders
```
Grammar nit - don't start a sentence with "With".

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
-with the language to understand, and for somebody familiar with the compiler to implement.
-This should get into specifics and corner-cases, and include examples of how the feature is used.
+With field defaults a caller can initialise a struct with default values without needing builders
+or a constructor function:
```
This paragraph should be in the motivation section. The design section should start with the nuts and bolts, probably syntax and then an example.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
+    c: 0,
+    ..Foo {
+        b: false
+    }
```
Note that this is not valid syntax - the struct in the FRU clause must be a proper struct object.

KodrAus replied:
Ah good point, I'll fix that up

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
-with the language to understand, and for somebody familiar with the compiler to implement.
-This should get into specifics and corner-cases, and include examples of how the feature is used.
+With field defaults a caller can initialise a struct with default values without needing builders
+or a constructor function:
```
Yeah, I would start by covering the syntax precisely, maybe even add a formal grammar. I would then try and summarise the semantics in a paragraph, before getting into the examples and edge cases (which are good and you should keep).

KodrAus replied:
Good idea. I'll spend some time with [the grammar spec](https://doc.rust-lang.org/grammar.html) and put one together. This might surface some inconsistencies too.

KodrAus replied:
Ok so I've gone and added what I think is a reasonable grammar, but am not sure if it's correct. I've only included the c-like structs, because that's all that's supported.

Does this look reasonable to you @nrc?

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
+This means there are values that can't be used as defaults, such as a `Vec` with 3 elements.
+
+Allowing functionality to be injected into data initialisation through abuse of `Default` means struct literal initialisation 
+is no longer guaranteed to be pure.
```
Can you explain why this is bad?

KodrAus replied:
I need to think about this more before I can clarify it better. I like that Rust separates data from functionality, it makes things easy to reason about and compose. Allowing arbitrary functionality in the form of `default()` that isn't really about initialisation to sit with the data feels like it messes with that strict separation.

That's all quite fluffy, so I need to focus it better. Do you have any thoughts @nrc ?

nrc replied:
I would think in terms of user expectations - we expect literal expressions to have small, predictable runtime cost and to be totally deterministic (and trivially predictable).

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
 
-Why should we *not* do this?
+Field defaults are limited to `const` expressions and calls to `default()`.
+This means there are values that can't be used as defaults, such as a `Vec` with 3 elements.
```
I would give a more general example, such as "any value that requires allocation, including non-empty `Vec`s"

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
+its addition can't be treated as a non-breaking change.
+
+The goal of this design is to let users build a struct as if its default fields weren't there.
+
```
You should add type inference as an alternative too, I think

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
-# Unresolved questions
-[unresolved]: #unresolved-questions
+Allowing arbitrary expressions as defaults would make this feature more powerful, 
+but at the expense of allowing functionality to leak into the struct's data.
```
This sentence doesn't quite make sense.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
+Ultimately, the combination of `const`s and `Default` strikes the best balance between expressiveness of allowable
+field default values and constraints on their cost.
+When using `const` values, there is no runtime overhead for default fields.
+When using `Default`, the expected overhead is small.
```
I would add removing `default` as an alternative too.

KodrAus replied:
Sans-default is now what the RFC is propising, and `default` is in alternatives.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,206 @@
+This would make it clearer to callers that additional code could be run on struct initialisation,
+weakening arguments against more powerful default expressions.
+However it would prevent field default from being used to maintain backwards compatibility,
+and reduce overall ergonomics.
```
Could you add the back-compat issue to the motivation please?

nrc commented:
General comments: will this apply to struct variants? Tuple structs?

Could you wrap your lines at 80 columns and not use one sentence per line please? Many people read RFCs by the source (since it is the only way to comment), and it is quite hard to read atm.

leoyvens commented:
@KodrAus I think explicit type is the best thing to propose at least initially, but inference should be listed as an alternative.

Commit:
remove manual wrapping

KodrAus commented:
Thanks @nrc and @leodasvacas :+1: I'll work through this new round of feedback over the next day or two.

KodrAus commented:
@nrc RE Tuple structs, I haven't considered those yet, but the `..` syntax doesn't support them so I wouldn't expect field defaults to either.

KodrAus committed:
address some feedback

nrc commented:
@KodrAus yeah, I wouldn't expect this to work for tuple structs, but it is worth mentioning in the RFC.

KodrAus committed:
rework and move default to alternatives

KodrAus committed:
work on intro para

KodrAus committed:
rework grammar

KodrAus committed:
speell chek

leoyvens commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,207 @@
 [motivation]: #motivation
 
-Why are we doing this? What use cases does it support? What is the expected outcome?
+Field default address two issues with structure initialisation in Rust: privacy and boilerplate. This is achieved by letting callers omit fields from initialisation when a default is specified for that field. This syntax also allows structures to be updated in a non-breaking fashion, by providing defaults for new fields.
```
"updated" is a bit ambiguous here because of the term "functional record update", maybe just say "allows fields to be added to structs in a backwards compatible way"

KodrAus commented:
I've spent some time on this this morning. RFCs are a lot of work! I can't imagine how much effort must have gone into some of the bigger ones like MIR.

I'll give it another go over with fresh eyes this afternoon.

leoyvens commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,207 @@
+
+With the `..` syntax, values for missing fields can be taken from another structure. However, this still requires an already initialised struct after the `..`. It also isn't valid if the struct has innaccessible private fields.
+
+To work around these shortcomings, users can create constructor functions or more elaborate builders. The problem with a constructor is that you either need one for each combination of fields a caller can supply. Builders enable more advanced initialisation, but need additional boilerplate.
```
"either" is not parsing here. Also, an alternative for constructors is providing one constructor with the mandatory fields (those without default values), and expect the user to set the other fields after initialization. The fields have to be visible.

KodrAus replied:
Yeh @nrc didn't seem convinced that's something people actually do in a comment above, and I haven't personally done it either.

Is it really that common to initialise a struct with a constructor and then go and manually update the fields you want different?

leoyvens replied:
I have never seen an API that works like that, it's more of a hypotetical possibility, maybe not worthy of inclusion.

leoyvens commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,207 @@
+
+Field defaults could require callers to use an opt-in syntax like `..`. This would make it clearer to callers that additional code could be run on struct initialisation, weakening arguments against more powerful default expressions. However it would prevent field default from being used to maintain backwards compatibility, and reduce overall ergonomics.
+
+If it's unclear whether or not a particular caller will use a default field value then its addition can't be treated as a non-breaking change.
```
Maybe mention the private field trick to force users to finish their initialisers with `..`

KodrAus replied:
Is that about forcing that feature into all struct initialisers so new fields don't mean broken builds?

leoyvens replied:
It's about if I want adding fields to my struct to be backwards compatible, I can add a private field that does nothing, therefore forcing users of my struct to intialize it with `..`.

KodrAus replied:
Yup, gotcha 👍 I'll make a note of it

KodrAus committed:
fix a few syntax issues

KodrAus committed:
add examples for validity

KodrAus committed:
add example for private defaults

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
 [summary]: #summary
 
-One para explanation of the feature.
+Allow structure definitions to supply default values for individual fields, and then allow those fields to be omitted from struct initialisation:
```
s/structure/struct

nrc replied:
and elsewhere

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
 [motivation]: #motivation
 
-Why are we doing this? What use cases does it support? What is the expected outcome?
+Field defaults address two issues with structure initialisation in Rust: privacy and boilerplate. This is achieved by letting callers omit fields from initialisation when a default is specified for that field. This syntax also allows allows fields to be added to structures in a backwards compatible way, by providing defaults for new fields.
```
I would take out this description of the proposal and keep the motivation for a description of the problems

KodrAus replied:
This description was meant to be the summary for the motivation. I'll rework it a bit.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
-Why are we doing this? What use cases does it support? What is the expected outcome?
+Field defaults address two issues with structure initialisation in Rust: privacy and boilerplate. This is achieved by letting callers omit fields from initialisation when a default is specified for that field. This syntax also allows allows fields to be added to structures in a backwards compatible way, by providing defaults for new fields.
+
+Rust allows you to create an instance of a structure using a literal syntax. This requires all fields in the structure be assigned a value, and can't be used with private fields. It can also be inconvenient for large structures whose fields usually receive the same values.
```
You should explain why private fields cause problems

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
+
+Rust allows you to create an instance of a structure using a literal syntax. This requires all fields in the structure be assigned a value, and can't be used with private fields. It can also be inconvenient for large structures whose fields usually receive the same values.
+
+With the `..` syntax, values for missing fields can be taken from another structure. However, this still requires an already initialised struct after the `..`. It also isn't valid if the struct has innaccessible private fields.
```
I would name the syntax as functional record update here (even though it is a terrible name)

nrc replied:
Are you sure about the private fields thing? Maybe this statement needs some qualification (or at least explanation)

KodrAus replied:
It seems to be the case, according to [this gist](http://play.integer32.com/?gist=16da2ae9f6a8220c5530ceafdd1a5264).

If there are any private fields then the functional record update borks. But that seems to go against what @leodasvacas is saying about using private fields to force callers to use it.

leoyvens replied:
@KodrAus The private field trick would be a hypothetical trick if we used explicit syntax for default values. Currently it dosent work because FRU and private fields dont go along.

KodrAus replied:
Ah gotcha :)

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
+
+With the `..` syntax, values for missing fields can be taken from another structure. However, this still requires an already initialised struct after the `..`. It also isn't valid if the struct has innaccessible private fields.
+
+To work around these shortcomings, users can create constructor functions or more elaborate builders. The problem with a constructor is that you need one for each combination of fields a caller can supply. Builders enable more advanced initialisation, but need additional boilerplate.
```
I would give some examples here

KodrAus replied:
I added a really trivial example of a constructor function and builder method

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
+
+To work around these shortcomings, users can create constructor functions or more elaborate builders. The problem with a constructor is that you need one for each combination of fields a caller can supply. Builders enable more advanced initialisation, but need additional boilerplate.
+
+Field defaults allow a caller to initialise a structure with default values without needing builders or a constructor function:
```
Either this bit should be removed or moved to the next section, or you should change it to directly answer the problems raised earlier in this section.

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,268 @@
+In the initialiser for a structure, a default value expression can be optionally supplied for a field:
+
+```
+struct_field : ident ':' type_path |
````
nit: I'd use `::=` instead of `:`

nrc replied:
Oh, you should have an optional `pub` in there

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,268 @@
+               ident ':' type_path '=' expr
+```
+
+The syntax is modeled after `const` expressions. Field defaults are only valid for classic C structures:
````
This should be two paras, I don't think you need the second grammar and better to explicitly say that you don't support tuple structs.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
+    a: &'static str,
+    b: Vec<bool> = Vec::new(),
+                   ^^^^^^^^^^
+                   // error: calls in field defaults are limited to struct and enum constructors
```
Is this error message adapted from consts?

KodrAus replied:
Yeh it is. I thought I'd include a bit of detail about errors, but also wasn't sure if it was necessary.

nrc replied:
Cool, I think it is nice to have this in, I'd note that the message is taken from consts and that as we extend the scope of constant expressions, this message will change.

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,268 @@
+}
+```
+
+Field defaults are sugar for the 'real' initialiser, where values for missing fields are added with the supplied default expression. 
````
It is not really sugar, since it must be done much later in the compiler than syntax desugaring.

KodrAus replied:
Ah fair enough. Is there a technical term I can use for interpreting and replacing this with a complete initialisation? How does function record update work?

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,268 @@
+};
+```
+
+The field default should only be used when the caller doesn't supply a value, to avoid unnecessarily assigning values.
````
You could cut this para

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,268 @@
+
+// `b` is `true`, even though `bool::default()` is `false`
+let foo = Foo::default();
+```
````
You should note that this means `derive(Default)` can be used more widely because only fields with default values need to implement `Default`

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
 [drawbacks]: #drawbacks
 
-Why should we *not* do this?
+Field defaults are limited to `const` expressions. This means there are values that can't be used as defaults, such any value that requires allocation, including common collections like `Vec`.
```
typo: "such any"

nrc replied:
and I think you meant `Vec::new()` not just `Vec`

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
 
-# Unresolved questions
-[unresolved]: #unresolved-questions
+Allowing arbitrary expressions as field defaults would make this feature more powerful. However, limiting field defaults to `const`s maintains the expectation that struct literals are cheap and deterministic. The same isn't true when arbitrary expressions that could reasonably panic or block on io are allowed.
```
s/consts/constant expressions

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,268 @@
-[unresolved]: #unresolved-questions
+Allowing arbitrary expressions as field defaults would make this feature more powerful. However, limiting field defaults to `const`s maintains the expectation that struct literals are cheap and deterministic. The same isn't true when arbitrary expressions that could reasonably panic or block on io are allowed.
+
+For complex initialisation logic, builders are the preferred option because they don't need to carry this same expectation.
```
s/need to//

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,268 @@
+```rust
+struct Foo {
+    a: &'static str,
+    b = 42,
````
Integers are not the best example for this because of the more complex type inference, maybe better to use a string or a struct.

nrc commented:
@KodrAus looking really good, I have another round of comments, but I think they are all polishing now. Nearly done :-)

KodrAus committed:
address some feedback

leoyvens commented:
The proposal in rust-lang/rfcs#448 seems to overlap with this, at least syntactically. ping @comex

KodrAus committed:
address some more feedback

KodrAus committed:
add ctor and builder example

KodrAus committed:
fix typo in grammar

KodrAus committed:
keep flow through examples

KodrAus committed:
clarify errors

KodrAus committed:
clarify default derive

KodrAus committed:
rename examples section

KodrAus committed:
add trailing comma

KodrAus commented:
@nrc @leodasvacas I've worked through this round of feedback. I'm not totally convinced on the constructor / builder example  I have though. It adds a bit of context but is pretty trivial.

Any thoughts?

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,291 @@
+    fn b(mut self, b: bool) -> Self {
+        self.b = b;
+        self
+    }
```
The usual builder pattern has a `FooBuilder` struct and methods on that, rather than on `Foo` itself. You could keep the ctor example, remove the builder example and instead refer to a builder in the std lib.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,291 @@
+}
+
+impl Foo {
+    /// Constructor function.
```
nit: I'd prefer regular comments to doc comments

KodrAus committed:
replace doc comments with comments

KodrAus committed:
remove builder example

KodrAus commented:
@nrc Sorted those little tweaks.

leoyvens commented:
[This is the RFC that made FRU work only for public fields](https://github.com/rust-lang/rfcs/pull/736). Seems like a relevant read.

KodrAus commented:
Thanks @leodasvacas. So this RFC contradicts this statement from the one you posted:

> If any fields are hidden by privacy, then all forms of struct literal syntax are unavailable, including FRU.)

It does solve the mentioned drawback of backwards compatibility though. And I think it does so in a good way. You don't need to derive Default and you don't have to default fields that shouldn't have a default value. So the one declaring the struct has full control over how it can be initialised.

nrc commented:
I'd note that the mechanism we're proposing here is fundamentally different in that the opt-in is in the struct def, not in the user of the struct. Put another way, if you want to ensure that a field is really, really private, then just don't give it a default.

leoyvens commented:
@KodrAus You are referring to the suggestion in the RFC of having a pub member that contains the private members? That is a nice workaround but most library authors don't seem to keep it in mind.

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,285 @@
+
+The problem with a constructor is that you need one for each combination of fields a caller can supply. To work around this, you can use builders, like [`process::Command`](https://doc.rust-lang.org/stable/std/process/struct.Command.html) in the standard library. Builders enable more advanced initialisation, but need additional boilerplate.
+
+This RFC proposes a solution to improve `struct` literal ergonomics so they can be used for `struct`s with private fields and to reduce initialisation boilerplate for simple scenarios. This is achieved by letting callers omit fields from initialisation when a default is specified for that field. This syntax also allows allows fields to be added to `struct`s in a backwards compatible way, by provi ...
```
nit: needs a comma after "ergonomics" and "private fields".

Typo: "allows allows"

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,285 @@
+let foo = Foo {
+    a: "Hello",
+    c: 42,
+};
```
I think it would be interesting for this example to give defaults for every field and then show a few different struct lits supplying different combinations of fields.

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,285 @@
+
+```
+struct_field ::= vis ? ident ':' type_path |
+                 vis ? ident ':' type_path '=' expr
````
nit: remove space between `vis` and `?`

nrc commented on default-fields/0000-default-fields.md:
````diff
@@ -6,31 +6,285 @@
+                 vis ? ident ':' type_path '=' expr
+```
+
+The syntax is modeled after constant expressions. Field defaults are only valid for classic C `struct`s, so tuple `struct`s aren't supported.
````
I would change the second sentence to "Field defaults for tuple structs are not supported."

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,285 @@
+
+## Interpretation
+
+The value of a field default must be a compile-time expression. So any expression that's valid as `const` can be used as a field default. This ensures values that aren't specified by the caller are deterministic and cheap to produce. A type doesn't need to derive `Default` to be valid as a field default.
```
"valid as _a_"

nrc commented on default-fields/0000-default-fields.md:
```diff
@@ -6,31 +6,285 @@
+    a: "Hello",
+    b: true,
+    c: 42,
+};
```
It's probably worth noting here that it is not quite a shorthand, because if `b` were private, then this wouldn't work.

nrc commented:
Looking good!

KodrAus committed:
work on some feedback

KodrAus committed:
add extra note on field privacy

KodrAus commented:
Sorry for a big delay once again! I decided to do a workshop this month so had to learn `tokio` :)

nrc commented:
Merged! I think this is looking good and is ready to submit as an RFC PR.

@KodrAus are you happy to submit it? Any questions about the process there?
